src/Microsoft.OData.Edm/Csdl/Parsing/SchemaJsonParser.cs

```diff
@@ -655,13 +655,20 @@ internal static CsdlKey ParseCsdlKey(string name, JsonElement keyArray, JsonPars
             {
                 if (v.ValueKind == JsonValueKind.Object)
                 {
-                    // TODO: ODL doesn't support the key referencing a property of a complex type as below.
                     // "$Key": [
                     //    {
                     //      "EntityInfoID": "Info/ID"
                     //    }
                     //  ]
-                    p.ReportError(EdmErrorCode.InvalidKeyValue, "It doesn't support the key object.");
+                    string pName = "";
+                    string pValue = "";
+                    v.ParseAsObject(p, (keyAlias, keyName) =>
+                    {
+                        pName = keyName.ParseAsString(p);
+                        pValue = keyAlias;
+                    });
+
+                    return new CsdlPropertyReference(pName, pValue, context.Location());

```

Comment:
Is there any other scenario where we reach this branch? The `ReportError` message doesn't seem specific to this specific feature (even if the comment is specific to the feature). Maybe we should be more a little bit more defensive, something like:

```suggestion
                    bool error = false;
                    string pName = null;
                    string pValue = null;
                    v.ParseAsObject(p, (keyAlias, keyName) =>
                    {
                        if (pName != null || pValue != null)
                        {
                            error = true;
                            return;
                        }
                        
                        pName = keyName.ParseAsString(p);
                        pValue = keyAlias;
                    });
                    if (error || pName == null || pValue == null)
                    {
                        p.ReportError(EdmErrorCode.InvalidKeyValue, "It doesn't support the key object.");
                    }
                    else
                    {
                        return new CsdlPropertyReference(pName, pValue, context.Location());
                    }
```

Alternatively, we could do some `Debug.Assert`s if we want to be a little more flexible, but that might actually result in bigger issues for customers if something goes wrong.

Reply:
so before we were returning the error because the feature was not supported. Now that the feature is being supported, there is no need to return the error.

Reply:
I guess I'm just making sure that the *only* time we return that error is because the feature was not supported. Is that right?

Reply:
Since the feature *is* supported, I think that error message would be misleading. If we want to validate against null key or value, I think the error message should be adjusted accordingly.

Reply:
the key cannot be null here.